bridges/snowbridge/primitives/router/src/inbound/v2.rs

```diff
@@ -0,0 +1,642 @@
+	ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+	GatewayProxyAddress: Get<H160>,
+	EthereumUniversalLocation: Get<InteriorLocation>,
+	GlobalAssetHubLocation: Get<Location>,
+{
+	_phantom: PhantomData<(
+		EthereumNetwork,
+		InboundQueuePalletInstance,
+		ConvertAssetId,
+		GatewayProxyAddress,
+		EthereumUniversalLocation,
+		GlobalAssetHubLocation,
+	)>,
+}
+
+impl<
+		EthereumNetwork,
+		InboundQueuePalletInstance,
+		ConvertAssetId,
+		GatewayProxyAddress,
+		EthereumUniversalLocation,
+		GlobalAssetHubLocation,
+	> ConvertMessage
+	for MessageToXcm<
+		EthereumNetwork,
+		InboundQueuePalletInstance,
+		ConvertAssetId,
+		GatewayProxyAddress,
+		EthereumUniversalLocation,
+		GlobalAssetHubLocation,
+	>
+where
+	EthereumNetwork: Get<NetworkId>,
+	InboundQueuePalletInstance: Get<u8>,
+	ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+	GatewayProxyAddress: Get<H160>,
+	EthereumUniversalLocation: Get<InteriorLocation>,
+	GlobalAssetHubLocation: Get<Location>,
+{
+	fn convert(
+		message: Message,
+		origin_account_location: Location,
+	) -> Result<(Xcm<()>, u128), ConvertMessageError> {
+		let mut message_xcm: Xcm<()> = Xcm::new();
+		if message.xcm.len() > 0 {
+			let xcm_string = hex::encode(message.xcm.clone());
+			log::info!(target: LOG_TARGET,"found xcm payload: {:x?}", xcm_string);
+			// Allow xcm decode failure so that assets can be trapped on AH instead of this
+			// message failing but funds are already locked on Ethereum.
+			if let Ok(versioned_xcm) = VersionedXcm::<()>::decode_with_depth_limit(
+				MAX_XCM_DECODE_DEPTH,
+				&mut message.xcm.as_ref(),
+			) {
+				if let Ok(decoded_xcm) = versioned_xcm.try_into() {
+					message_xcm = decoded_xcm;
+				} else {
+					log::error!(target: LOG_TARGET,"unable to decode xcm");
+				}
+			} else {
+				log::error!(target: LOG_TARGET,"unable to decode versioned xcm");
+			}
+		}
+
+		log::info!(target: LOG_TARGET,"xcm decoded as {:?}", message_xcm);
+
+		let network = EthereumNetwork::get();
+
+		// use eth as asset
+		let fee_asset = Location::new(2, [GlobalConsensus(EthereumNetwork::get())]);
+		let fee: XcmAsset = (fee_asset.clone(), message.execution_fee).into();
+		let eth: XcmAsset =
+			(fee_asset.clone(), message.execution_fee.saturating_add(message.value)).into();
+		let mut instructions = vec![
+			DescendOrigin(PalletInstance(InboundQueuePalletInstance::get()).into()),
+			UniversalOrigin(GlobalConsensus(network)),
+			ReserveAssetDeposited(eth.into()),
+			PayFees { asset: fee },
+		];
+		let mut reserve_assets = vec![];
+		let mut withdraw_assets = vec![];
+
+		let mut refund_surplus_to = origin_account_location;
+
+		if let Some(claimer) = message.claimer {
+			// If the claimer can be decoded, add it to the message. If the claimer decoding fails,
+			// do not add it to the message, because it will cause the xcm to fail.
+			if let Ok(claimer) = Junction::decode(&mut claimer.as_ref()) {
+				let claimer_location: Location = Location::new(0, [claimer.into()]);
```

Comment:
why is the claimer just a `Junction` / `InteriorLocation`? seems like a gratuitous limitation. You should leave it as full `Location` specifiable on Ethereum side.

Could be a local account (just a Junction like now) or it could be a remote location that will be derived into a sov acc.

Reply:
Addressed: https://github.com/vgeddes/polkadot-sdk/blob/snowbridge-v2/bridges/snowbridge/primitives/inbound-queue/src/v2/converter.rs#L121